### PR TITLE
Print Progress

### PR DESCRIPTION
**Pull request by billyjbryant**
### Description

This PR is an addendum to #52 and adds printer status and progress info to the Home Tab

image

### CODE REVIEW

**Comment by chrivers** [comment]
This is very exciting! :)

There's one thing I think that detracts a lot from good work you put in, and it's the fact that all the right-hand sides of the box pairs (elapsed, remain, layer, speed, etc) are all different widths. Same goes for the two top boxes with icons.

Isn't there some way to make this a little more controlled?

I really like the idea.

**Comment by thomasjpatterson** [comment]
@chrivers If I'm not mistaken that's a bootstrap thing so unless we jumped to a different underlying framework for the CSS that's the way it'll be.

**Comment by billyjbryant** [comment]
> There's one thing I think that detracts a lot from good work you put in, and it's the fact that all the right-hand sides of the box pairs (elapsed, remain, layer, speed, etc) are all different widths. Same goes for the two top boxes with icons.


> @chrivers If I'm not mistaken that's a bootstrap thing so unless we jumped to a different underlying framework for the CSS that's the way it'll be.

Yes, @chrivers , @thomasjpatterson is correct, because I am using bootstrap it is just part of that and determines button size based on content. In this case I'm using a form-group with a Button as a label and a disabled input field to display the content. This presents a nice interface but I do agree the sizing of the labels might be a tad distracting. I will see what I can do to fix that or I may call upon @just-trey's expertise here to see if he can improve upon it :)

**Comment by just-trey** [comment]
My gut for now tells me to leave it as is and maybe migrate away from bootstrap in a future release.

**Comment by billyjbryant** [comment]
I tweaked it a bit but by doing so it is no longer responsive. 
image

**Comment by billyjbryant** [comment]
> @billyjbryant  why are you doing the UI like that - why not use:
> https://getbootstrap.com/docs/5.3/forms/input-group/

I'm not quite sure what you mean because that's what I'm using... Did you mean this to be a quoted reply on a specific area of code?

File: static/ankersrv.js [hunk 1]
```diff
@@ -50,12 +50,107 @@ $(function () {
...
+        return weight;
+    }
+
     /**
      * Opens a websocket connection and outputs any incoming message data to console
      */
     socket = new WebSocket("ws://" + location.host + "/ws/mqtt");
     socket.addEventListener("message", (ev) => {
-        console.log(JSON.parse(ev.data));
+        const data = JSON.parse(ev.data);
```
Comment (chrivers) [inline static/ankersrv.js:122]: We are doing a lot of this same kind of mqtt parsing in the backend, for the upcoming mainsail branch.

In the short term, I'm ok with also doing the work in the frontend, as long as it's not much bigger than this, but if we plan to parse more messages like this, we should *really* consider importing the new parsing code meant for mainsail, so we don't double the efforts.

File: static/img/bed-icon.svg
Comment (chrivers) [inline static/img/bed-icon.svg:1]: What's the origin of these icons? Did you make them, or are they copied from some known source? This needs to be documented.

What's the license of the source material?

File: static/ankersrv.css [hunk 2]
```diff
@@ -1,41 +1,79 @@
...
+    display: inline-block;
+    position: fixed;
+    height: 25px;
+    width: 25px;
+}
+
+.icon-hotend {
+    background-color: #fff;
+    -webkit-mask-image: url(../static/img/hotend-icon.svg);
+    mask-image: url(../static/img/hotend-icon.svg);
```
Comment (chrivers) [inline static/ankersrv.css:62]: Wait, why `../static`?

How about just `img/hotend-icon.svg`?

File: static/ankersrv.css [hunk 3]
```diff
@@ -1,41 +1,79 @@
...
+}
+
+.progress {
+    height: 40px;
+}
+
+.icon {
+    background-size: cover;
+    display: inline-block;
+    position: fixed;
```
Comment (chrivers) [inline static/ankersrv.css:54]: Why are we using `position: fixed`, if we are not specifying the position?

File: static/index.html [hunk 4]
```diff
@@ -131,21 +132,240 @@ <h4>{{ printer_serial }}</h4>
...
+                                        <div class="col-lg-12 px-3 rounded-3">
+                                            <div class="row mt-3">
+                                                <h4>Temperature</h4>
+                                            </div>
+                                            <div class="row mt-2">
+                                                <div class="col-6">
+                                                    <div class="input-group">
+                                                        <div class="input-group-prepend print-temp-label">
+                                                            <div class="input-group-text mx-auto" id="nozzleGroup">
+                                                                &nbsp;<i class="icon-hotend icon"></i>
```
Comment (chrivers) [inline static/index.html:175]: Since `.icon` is our own css class, can't we add some spacing attribute to it, instead of this `&nbsp;` 90's layout hack? :smile:

**Comment by LazeMSS** [comment]
> > @billyjbryant  why are you doing the UI like that - why not use:
> > https://getbootstrap.com/docs/5.3/forms/input-group/
> 
> I'm not quite sure what you mean because that's what I'm using... Did you mean this to be a quoted reply on a specific area of code?

Basicaly all the input groups etc: https://github.com/billyjbryant/ankermake-m5-protocol/blob/billybryant/print_progress/static/index.html#L172

I will try and do a checkout and PR on this asap

File: static/ankersrv.js [hunk 5]
```diff
@@ -50,12 +50,107 @@ $(function () {
...
+        return Math.round(temp / 100);
+    }
+
+    /**
+     * Calculate the percentage between two numbers
+     * @param {number} layer
+     * @param {number} total
+     * @returns {number} percentage
+     */
+    function getPercentage(progress) {
```
[the remark this answers is not in the document]
Comment (LazeMSS) [reply static/ankersrv.js:68]: const progress = getPercentage(data.progress);

File: static/ankersrv.js [hunk 6]
```diff
@@ -50,12 +50,107 @@ $(function () {
...
+        return Math.round(((progress / 100) * 100) / 100);
+    }
+
+    /**
+     * Truncate a string to a specific length
+     * @param {string} str
+     * @param {number} maxLength
+     * @returns {string} Truncated string
+     */
+    function truncStr(str, maxLength) {
```
[the remark this answers is not in the document]
Comment (LazeMSS) [reply static/ankersrv.js:78]: And why not use .substring(0,length); and the use this class for the divs: https://getbootstrap.com/docs/5.3/helpers/text-truncation/

File: static/ankersrv.js (hunk 1 above)
Comment (LazeMSS) [reply static/ankersrv.js:122]: Or maybe just less if states - it could be made easier to ready at least or more template friendly:
https://github.com/LazeMSS/ankermake-m5-protocol/blob/9b867cd6bac90ac5b00c25c95794d11ee9732326/static/ankersrv.js#L38

File: static/img/bed-icon.svg
Comment (billyjbryant) [reply static/img/bed-icon.svg:1]: I will look up the license, they came from a royalty free svg site, I'll make sure I get the proper license in place.

Comment (LazeMSS) [reply static/img/bed-icon.svg:1]: https://www.iconfinder.com/iconsets/3d-printing-line
license: https://creativecommons.org/licenses/by/3.0/

File: static/ankersrv.css (hunk 3 above)
Comment (billyjbryant) [reply static/ankersrv.css:54]: This prevents the icon from tiling. If someone like @just-trey has a better way to do this, I'm open to hear it :)

File: static/index.html (hunk 4 above)
Comment (LazeMSS) [reply static/index.html:175]: Yes basically these are the input groups I was writing about: 
https://getbootstrap.com/docs/5.3/forms/input-group/

File: static/ankersrv.css (hunk 3 above)
Comment (thomasjpatterson) [reply static/ankersrv.css:54]: You are correct this is the easiest way to keep an image from trying to tile

**Comment by LazeMSS** [comment]
In general spacing/padding/margin:
https://getbootstrap.com/docs/5.3/utilities/spacing/ 

Spinners:
there is native css based spinners for that: https://getbootstrap.com/docs/5.3/components/spinners/ - avoids using images.

Generic css :
- !important should be a final resort
- text-decoration: none -> https://getbootstrap.com/docs/5.3/utilities/text/#text-decoration
- font-size: https://getbootstrap.com/docs/5.3/utilities/text/#font-size
- font-weight: https://getbootstrap.com/docs/5.3/utilities/text/#font-weight-and-italics

Basically there is "classes" for almost anything in bootstrap - that makes is more easier to maintain and does not break on some browsers etc.

File: static/index.html (hunk 4 above)
Comment (billyjbryant) [reply static/index.html:175]: The hack was actually to force the height of the bootstrap label, spacing wasn't a concern. Every time I took it out, no matter which bootstrap height class I used, the label was smaller than the height of the icon.

Comment (thomasjpatterson) [reply static/index.html:175]: you'd be surprised by the amount of `@nbsp;` in code to get it to move over a space.

File: static/ankersrv.css (hunk 3 above)
Comment (LazeMSS) [reply static/ankersrv.css:54]: why not use
background-repeat: no-repeat

File: static/img/bed-icon.svg
Comment (billyjbryant) [reply static/img/bed-icon.svg:1]: Thank you for the assist @LazeMSS, I'll update the PR to add proper attribution.

**Comment by billyjbryant** [comment]
@LazeMSS 

> Spinners: there is native css based spinners for that: https://getbootstrap.com/docs/5.3/components/spinners/ - avoids using images.

I'm not sure what you are referring to regarding a spinner? If you are mentioning the `poster` attribute of the HTML5 `<video>` tagset, how would you suggest implementing a spinner here? 


> !important should be a final resort

I always use `!important` as a final resort. Bootstrap's CSS, despite changing the ordering in the html, is loading after the `ankerctl.css` file. In order to override certain aspects of Bootstrap, I was relegated to using `!important`, if you can find another way to do this, by all means, you are more than welcome to update it to function properly without using `!important`. 

> text-decoration: none -> https://getbootstrap.com/docs/5.3/utilities/text/#text-decoration
font-size: https://getbootstrap.com/docs/5.3/utilities/text/#font-size
font-weight: https://getbootstrap.com/docs/5.3/utilities/text/#font-weight-and-italics

I'm more than happy to shift to these, I will say; however, that the bootstrap classes for text-decoration were not functioning properly on the ONLY link that I implemented the `text-decoration:none` style attribute for.

File: static/ankersrv.js [hunk 7]
```diff
@@ -50,12 +50,107 @@ $(function () {
...
+    }
+
     /**
      * Opens a websocket connection and outputs any incoming message data to console
      */
     socket = new WebSocket("ws://" + location.host + "/ws/mqtt");
     socket.addEventListener("message", (ev) => {
-        console.log(JSON.parse(ev.data));
+        const data = JSON.parse(ev.data);
+        if (data.commandType == 1001) {
```
Comment (chrivers) [inline static/ankersrv.js:123]: Since we have `.stl` files for the protocol specification, we can easily generate some a `.js` file with symbolic names for these MQTT constants.

This would make it a lot easier to follow what the decode logic is here.

I've pushed a new branch that demonstrates this. (https://github.com/Ankermgmt/ankermake-m5-protocol/blob/chrivers/basic-libflagship-js-support/static/libflagship.js)

If that works, I'll get it merged into the upcoming branch.

File: static/index.html [hunk 8]
```diff
@@ -240,7 +245,10 @@ <h4>Print Progress</h4>
                                                     <div class="col-6">
                                                         <div class="input-group">
                                                             <div class="input-group-prepend">
-                                                                <div class="input-group-text" id="timeElapsed">
+                                                                <div
+                                                                    class="input-group-text print-info-label"
```
Comment (chrivers) [inline static/index.html:249]: We're adding this specific class to a number of elements here, and they all seem to be selectable by a rule such as `input-group-prepend input-group-text` or similar.

Do we really need a new css class here?

**Comment by LazeMSS** [comment]
> @LazeMSS
> 
> > Spinners: there is native css based spinners for that: https://getbootstrap.com/docs/5.3/components/spinners/ - avoids using images.
> 
> I'm not sure what you are referring to regarding a spinner? If you are mentioning the `poster` attribute of the HTML5 `<video>` tagset, how would you suggest implementing a spinner here?

I would implement a div that contains the "Please wait" in text and with a css spinner. Make the video layer hidden until the video play event happens (https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play_event) then hide the loading div and show the video layer.

> 
> > !important should be a final resort
> 
> I always use `!important` as a final resort. Bootstrap's CSS, despite changing the ordering in the html, is loading after the `ankerctl.css` file. In order to override certain aspects of Bootstrap, I was relegated to using `!important`, if you can find another way to do this, by all means, you are more than welcome to update it to function properly without using `!important`.
> 
I would still say a need to use important is breaking - so if we switch to a new bootstrap theme/color (like add support for dark/normal mode) then these colors are forced instead of using the colors defined in the scheme.

> > text-decoration: none -> https://getbootstrap.com/docs/5.3/utilities/text/#text-decoration
> > font-size: https://getbootstrap.com/docs/5.3/utilities/text/#font-size
> > font-weight: https://getbootstrap.com/docs/5.3/utilities/text/#font-weight-and-italics
> 
> I'm more than happy to shift to these, I will say; however, that the bootstrap classes for text-decoration were not functioning properly on the ONLY link that I implemented the `text-decoration:none` style attribute for.

I don't even understand why the bottom link "View on github" is  btn-link class - why not just a normal link?

File: static/ankersrv.css (hunk 3 above)
Comment (just-trey) [reply static/ankersrv.css]: @chrivers, this is fixed with my push. also, you will see new clock icons added are custom-built by me. What I found online was too fat.

File: static/ankersrv.css (hunk 2 above)
Comment (just-trey) [reply static/ankersrv.css]: Stacik is where all the static resources are. Do we need that changed? The index and CSS are in there too.

File: static/index.html (hunk 4 above)
Comment (just-trey) [reply static/index.html]: &sbsp is now gone with the latest fixes.

File: static/img/bed-icon.svg
Comment (chrivers) [reply static/img/bed-icon.svg:1]: @billyjbryant Thanks. This is a MUSTFIX, since otherwise it's a violation of the upstream license.

File: static/index.html (hunk 4 above)
Comment (chrivers) [reply static/index.html]: @just-trey awesome, thanks :)

File: static/img/load-screen.svg
Comment (chrivers) [inline static/img/load-screen.svg:1]: This is super cool, @just-trey. Nice work :+1:

File: static/img/bed-icon.svg
Comment (chrivers) [reply static/img/bed-icon.svg:1]: @just-trey confirmed that the new icons are his original work, and so there's no upstream license to worry about.

File: static/ankersrv.css (hunk 3 above)
Comment (chrivers) [reply static/ankersrv.css]: Looks great, thanks.

File: static/ankersrv.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrivers) [reply static/ankersrv.js:68]: I'll add a couple of commits after merging this.

File: static/index.html (hunk 8 above)
Comment (billyjbryant) [reply static/index.html]: > We're adding this specific class to a number of elements here, and they all seem to be selectable by a rule such as `input-group-prepend input-group-text` or similar.
> 
> Do we really need a new css class here?

I did this so that we only effected these fields and not any other field also using the `input-group-text` class (which is used on the Upload form for the config file.

File: static/ankersrv.css [hunk 9]
```diff
@@ -1,41 +1,79 @@
...
+
+.icon-bed {
+    background-color: #fff;
+    -webkit-mask-image: url(../static/img/bed-icon.svg);
+    mask-image: url(../static/img/bed-icon.svg);
+}
+
+.print-info-label {
+    width: 75px;
+    padding-left: 5px;
```
[the remark this answers is not in the document]
Comment (billyjbryant) [reply static/ankersrv.css]: This was a forced fix to make the labels a set width, I tried using bootstrap classes and it wasn't working properly. This addresses your original comment regarding the labels being different sizes across the elements.

File: static/ankersrv.css [hunk 10]
```diff
@@ -1,41 +1,79 @@
...
+.icon-hotend {
+    background-color: #fff;
+    -webkit-mask-image: url(../static/img/hotend-icon.svg);
+    mask-image: url(../static/img/hotend-icon.svg);
+}
+
+.icon-bed {
+    background-color: #fff;
+    -webkit-mask-image: url(../static/img/bed-icon.svg);
+    mask-image: url(../static/img/bed-icon.svg);
```
[the remark this answers is not in the document]
Comment (billyjbryant) [reply static/ankersrv.css]: They are black SVGs, I did this here so that I could override the color without having to modify the SVG itself and found no other good way to do this.

File: static/ankersrv.js [hunk 11]
```diff
@@ -50,12 +50,107 @@ $(function () {
...
+    }
+
+    /**
+     * Get the filament weight based on its length
+     * @param {number} length
+     * @returns {number} Weight
+     */
+    function getWeight(length) {
+        const diameter = 1.75;
+        const density = 1.24;
```
[the remark this answers is not in the document]
Comment (billyjbryant) [reply static/ankersrv.js:110]: This function actually needs to go away. I was originally using it to convert how much Filament was being used because the print I was doing was returning `mm` for the filament measurement but it didn't seem to be accurate so I stripped it from the visible fields.

File: static/ankersrv.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (chrivers) [reply static/ankersrv.css]: Ah, that makes sense. Thanks for the explanation :)

File: static/ankersrv.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (chrivers) [reply static/ankersrv.css]: Ah, it seems bootstrap has some weird quirks with the way sizing interacts with this. And with @just-trey's changes on top, I'd consider this resolved now :)

**Review by chrivers — APPROVED** [review]
Looks like we've addressed all conversations, so let's get this thing merged :)
